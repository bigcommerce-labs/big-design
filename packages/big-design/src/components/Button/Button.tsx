import React, { ComponentPropsWithoutRef, forwardRef, memo, Ref } from 'react';

import { MarginProps } from '../../mixins';
import { ProgressCircle } from '../ProgressCircle';

import { ContentWrapper, LoadingSpinnerWrapper, StyledButton } from './styled';

interface PrivateProps {
  forwardedRef: Ref<HTMLButtonElement>;
}

export interface ButtonProps extends ComponentPropsWithoutRef<'button'>, MarginProps {
  actionType?: 'normal' | 'destructive';
  children?: React.ReactNode;
  iconLeft?: React.ReactNode;
  iconOnly?: React.ReactNode;
  iconRight?: React.ReactNode;
  isLoading?: boolean;
  mobileWidth?: 'auto' | '100%';
  variant?: 'primary' | 'secondary' | 'subtle' | 'utility';
}

const LoadingSpinner = () => (
  <LoadingSpinnerWrapper alignItems="center">
    <ProgressCircle size="xxSmall" />
  </LoadingSpinnerWrapper>
);

const RawButton: React.FC<ButtonProps & PrivateProps> = memo(
  ({ forwardedRef, isLoading, disabled, ...props }) => {
    return (
      <StyledButton
        className="bd-button"
        {...props}
        disabled={isLoading || disabled}
        ref={forwardedRef}
      >
        {isLoading ? <LoadingSpinner /> : null}
        <ContentWrapper isLoading={isLoading}>
          {!props.iconOnly && props.iconLeft}
          {props.iconOnly}
          {!props.iconOnly && props.children}
          {!props.iconOnly && props.iconRight}
        </ContentWrapper>
      </StyledButton>
    );
  },
);

export const StyleableButton = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <RawButton {...props} forwardedRef={ref} />
));

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, style, ...props }, ref) => <RawButton {...props} forwardedRef={ref} />,
);

const defaultProps = {
  actionType: 'normal' as const,
  isLoading: false,
  mobileWidth: '100%' as const,
  variant: 'primary' as const,
};

Button.displayName = 'Button';
Button.defaultProps = { ...defaultProps };

StyleableButton.displayName = 'StyleableButton';
StyleableButton.defaultProps = { ...defaultProps };
