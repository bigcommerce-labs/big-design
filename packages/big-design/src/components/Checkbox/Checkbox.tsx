import { CheckIcon, RemoveIcon } from '@bigcommerce/big-design-icons';
import React, { forwardRef, useMemo, Ref } from 'react';

import { typedMemo, warning } from '../../utils';
import { useUniqueId } from '../../utils/useUniqueId';

import { CheckboxContainer, HiddenCheckbox, StyledCheckbox } from './styled';
import { CheckboxLabel } from './Label';

interface Props {
  hiddenLabel?: boolean;
  isIndeterminate?: boolean;
  label: React.ReactChild;
}

interface PrivateProps {
  forwardedRef: Ref<HTMLInputElement>;
}

export type CheckboxProps = Props & React.InputHTMLAttributes<HTMLInputElement>;

const RawCheckbox: React.FC<CheckboxProps & PrivateProps> = ({
  checked,
  className,
  disabled,
  hiddenLabel,
  isIndeterminate,
  label,
  forwardedRef,
  style,
  ...props
}) => {
  const id = props.id ? props.id : useUniqueId('checkbox');
  const labelId = useUniqueId('checkbox_label');

  const renderedLabel = useMemo(() => {
    if (!label) {
      return null;
    }

    if (typeof label === 'string') {
      return (
        <CheckboxLabel disabled={disabled} hidden={hiddenLabel} htmlFor={id} aria-hidden={disabled} id={labelId}>
          {label}
        </CheckboxLabel>
      );
    }

    if (React.isValidElement(label) && label.type === CheckboxLabel) {
      return React.cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, {
        hidden: hiddenLabel,
        htmlFor: id,
        id: labelId,
      });
    }

    warning('label must be either a string or a CheckboxLabel component.');
  }, [disabled, hiddenLabel, id, label, labelId]);

  return (
    <CheckboxContainer className={className} style={style}>
      <HiddenCheckbox
        type="checkbox"
        checked={checked}
        id={id}
        disabled={disabled}
        {...props}
        aria-labelledby={labelId}
        ref={checkbox => {
          if (checkbox && typeof isIndeterminate === 'boolean') {
            checkbox.indeterminate = !checked && isIndeterminate;
          }

          if (forwardedRef) {
            if (typeof forwardedRef === 'function') {
              forwardedRef(checkbox);
            } else {
              // RefObject.current is readonly in DefinitelyTyped
              // but in practice you can still write to it.
              // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31065
              (forwardedRef as React.MutableRefObject<HTMLInputElement | null>).current = checkbox;
            }
          }
        }}
      />

      <StyledCheckbox
        disabled={disabled}
        isIndeterminate={isIndeterminate}
        checked={checked}
        htmlFor={id}
        aria-hidden={true}
      >
        {!checked && isIndeterminate ? <RemoveIcon /> : <CheckIcon />}
      </StyledCheckbox>
      {renderedLabel}
    </CheckboxContainer>
  );
};

export const Checkbox = typedMemo(
  forwardRef<HTMLInputElement, CheckboxProps>(({ className, style, ...props }, ref) => (
    <RawCheckbox {...props} forwardedRef={ref} />
  )),
);
