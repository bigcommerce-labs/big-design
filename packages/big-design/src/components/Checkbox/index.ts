export { Checkbox, type CheckboxProps } from './Checkbox';
export { CheckboxLabel } from './Label';
