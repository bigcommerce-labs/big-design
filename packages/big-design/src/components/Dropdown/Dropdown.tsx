import { Placement } from 'popper.js';
import React, { AllHTMLAttributes } from 'react';
import { Manager, Reference, RefHandler } from 'react-popper';
import scrollIntoView from 'scroll-into-view-if-needed';

import { uniqueId } from '../../utils';
import { ListAction } from '../List/Action/Action';
import { ListItem } from '../List/Item/Item';
import { List } from '../List/List';

interface DropdownState {
  highlightedIndex: number;
  isOpen: boolean;
}

interface Props {
  maxHeight?: number;
  placement?: Placement;
  trigger: React.ReactElement;
  onActionClick?(): void;
  onItemClick?(value: AllHTMLAttributes<HTMLElement>['value']): void;
}

type DropdownProps = Props & React.HTMLAttributes<HTMLUListElement>;

export class Dropdown extends React.PureComponent<DropdownProps, DropdownState> {
  static Action = ListAction;
  static Item = ListItem;

  static readonly defaultProps: Partial<Props> = {
    placement: 'bottom',
  };

  state = {
    highlightedIndex: -1,
    isOpen: false,
  };

  private listRef: HTMLUListElement | null = null;
  private triggerRef: HTMLElement | null = null;

  private readonly uniqueDropdownId = uniqueId('dropdown_');
  private readonly uniqueTriggerId = uniqueId('trigger_');

  render() {
    const { children, maxHeight, onActionClick, onItemClick, placement, trigger, ...rest } = this.props;

    const highlightedItem = this.getHighlightedItem();
    const aria = highlightedItem ? { 'aria-activedescendant': highlightedItem.id } : {};

    return (
      <Manager>
        <Reference innerRef={node => (this.triggerRef = node)}>{({ ref }) => this.renderTrigger(ref)}</Reference>
        <List
          aria-labelledby={this.getTriggerId()}
          handleListRef={this.handleListRef}
          id={this.getDropdownId()}
          isOpen={this.state.isOpen}
          maxHeight={maxHeight}
          onKeyDown={this.handleOnDropdownKeyDown}
          placement={placement}
          role="menu"
          {...aria}
          {...rest}
        >
          {this.renderChildren()}
        </List>
      </Manager>
    );
  }

  private renderTrigger(ref: RefHandler) {
    const { trigger } = this.props;

    const aria = this.state.isOpen ? { 'aria-expanded': true, 'aria-owns': this.getDropdownId() } : {};

    return (
      React.isValidElement(trigger) &&
      React.cloneElement<React.HTMLAttributes<any> & React.RefAttributes<any>>(trigger, {
        'aria-haspopup': true,
        id: this.getTriggerId(),
        onClick: this.toggleList,
        ref,
        role: 'button',
        ...aria,
      })
    );
  }

  private renderChildren() {
    const { children } = this.props;

    return React.Children.map(
      children,
      (child, index) =>
        React.isValidElement(child) &&
        (child.type === ListItem || child.type === ListAction) &&
        React.cloneElement<React.LiHTMLAttributes<HTMLLIElement>>(child, {
          // @ts-ignore
          'data-highlighted': index === this.state.highlightedIndex,
          id: this.getItemId(child, index),
          onClick: child.type === ListItem ? this.handleOnItemClick : this.handleOnActionClick,
          onFocus: this.handleOnItemFocus,
          onMouseOver: this.handleOnItemMouseOver,
          role: 'menuitem',
        }),
    );
  }

  private toggleList = () => {
    this.state.isOpen ? this.closeList() : this.openList();
  };

  private openList() {
    this.setState({ isOpen: true }, () => {
      document.addEventListener('mousedown', this.handleOnClickOutside, false);

      // We need to wait until the menu is open to change the index
      this.setState({ highlightedIndex: 0 }, () => {
        return this.listRef && this.listRef.focus();
      });
    });
  }

  private closeList() {
    this.setState({ isOpen: false }, () => {
      document.removeEventListener('mousedown', this.handleOnClickOutside, false);

      return this.triggerRef && this.triggerRef.focus();
    });
  }

  private getDropdownId() {
    const { id } = this.props;

    return id || this.uniqueDropdownId;
  }

  private getItemId(item: React.ReactElement<React.LiHTMLAttributes<HTMLLIElement>>, index: number) {
    const { id } = item.props;

    return id || `${this.getDropdownId()}-item-${index}`;
  }

  private getTriggerId() {
    const { trigger } = this.props;
    const triggerId = trigger.props.id;

    return triggerId || this.uniqueTriggerId;
  }

  private findIndexByItem(item: HTMLLIElement) {
    const children = this.listRef && Array.from(this.listRef.children);

    return children && children.indexOf(item);
  }

  private getHighlightedItem() {
    return this.listRef && (this.listRef.children.item(this.state.highlightedIndex) as HTMLLIElement);
  }

  private getNumberOfChildren() {
    return React.Children.count(this.props.children);
  }

  private updateHighlightedIndex(highlightedIndex: number, cb?: () => void) {
    this.setState(
      {
        highlightedIndex,
      },
      () => {
        return cb && cb();
      },
    );
  }

  private handleListRef = (ref: HTMLElement | null) => {
    if (ref) {
      this.listRef = ref as HTMLUListElement;
    }
  };

  private handleOnActionClick = () => {
    const { onActionClick } = this.props;

    if (onActionClick) {
      onActionClick();
    }

    this.toggleList();
  };

  private handleOnClickOutside = (event: MouseEvent) => {
    if (!this.listRef || !this.triggerRef) {
      return;
    }

    if (this.listRef.contains(event.target as Node) || this.triggerRef.contains(event.target as Node)) {
      return;
    }

    this.toggleList();
  };

  private handleOnItemClick = () => {
    const { onItemClick } = this.props;
    const highlightedItem = this.getHighlightedItem();

    if (highlightedItem && onItemClick) {
      const value = highlightedItem.getAttribute('value');

      if (value) {
        onItemClick(value);
      }
    }

    this.toggleList();
  };

  private handleOnItemFocus = (event: React.FocusEvent<HTMLLIElement>) => {
    const index = this.findIndexByItem(event.currentTarget);

    return index !== null && this.updateHighlightedIndex(index);
  };

  private handleOnItemMouseOver = (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
    const index = this.findIndexByItem(event.currentTarget);

    return index !== null && this.updateHighlightedIndex(index);
  };

  /**
   * Accessiblilty Menu Keyboard Support
   * Learn more: https://www.w3.org/TR/wai-aria-practices/#menu
   */

  private handleOnDropdownKeyDown = (event: React.KeyboardEvent<HTMLUListElement>) => {
    const numberOfChildren = this.getNumberOfChildren();

    switch (event.key) {
      case 'Enter':
      case ' ': {
        event.preventDefault();
        this.handleOnItemClick();
        break;
      }
      case 'Escape': {
        event.preventDefault();
        event.stopPropagation();
        this.toggleList();
        break;
      }
      case 'ArrowUp':
      case 'ArrowLeft': {
        event.preventDefault();
        this.updateIndexAndScroll(this.nextItemIndex(-1));
        break;
      }
      case 'ArrowDown':
      case 'ArrowRight': {
        event.preventDefault();
        this.updateIndexAndScroll(this.nextItemIndex(1));
        break;
      }
      case 'Home': {
        event.preventDefault();
        this.updateIndexAndScroll(0);
        break;
      }
      case 'End': {
        event.preventDefault();
        this.updateIndexAndScroll(numberOfChildren - 1);
        break;
      }
      case 'Tab':
      case 'Esc':
      case 'Escape': {
        this.toggleList();
        break;
      }
    }
  };

  private nextItemIndex(direction: 1 | -1) {
    const nextIndex = this.state.highlightedIndex + direction;
    const numberOfChildren = this.getNumberOfChildren();

    if (nextIndex < 0) {
      return numberOfChildren - 1;
    } else if (nextIndex >= numberOfChildren) {
      return 0;
    } else {
      return nextIndex;
    }
  }

  private updateIndexAndScroll(index: number) {
    this.updateHighlightedIndex(index, this.scrollIntoView);
  }

  private scrollIntoView = () => {
    const element = this.getHighlightedItem();

    if (!element) {
      return;
    }

    return scrollIntoView(element, {
      behavior: 'smooth',
      scrollMode: 'if-needed',
    });
  };
}
