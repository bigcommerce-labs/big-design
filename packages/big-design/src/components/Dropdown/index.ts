export { Dropdown } from './Dropdown';

export {
  type DropdownItem,
  type DropdownItemGroup,
  type DropdownLinkItem,
  type DropdownProps,
} from './types';
