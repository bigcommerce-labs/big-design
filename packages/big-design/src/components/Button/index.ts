import { ButtonProps as _ButtonProps } from './Button';

export { Button } from './Button';
export type ButtonProps<T extends 'button' | 'a'> = _ButtonProps<T>;
