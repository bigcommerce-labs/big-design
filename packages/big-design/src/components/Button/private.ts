import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { ButtonProps, StyleableButton as _StyleableButton } from './Button';

export const StyleableButton = _StyleableButton;

export const MessagingButton = styled(_StyleableButton)<ButtonProps>`
  background-color: transparent;
  border: ${({ theme }) => theme.border.none};
  color: ${({ theme }) => theme.colors.secondary70};
  height: auto;
  padding: ${({ theme }) => theme.spacing.none};

  &:active {
    background-color: ${({ theme }) => theme.colors.primary20};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.primary20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.primary10};
  }

  &[disabled] {
    color: ${({ theme }) => theme.colors.secondary30};
  }
`;

MessagingButton.defaultProps = { theme: defaultTheme };
