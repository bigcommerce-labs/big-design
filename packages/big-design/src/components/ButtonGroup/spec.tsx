import { theme } from '@bigcommerce/big-design-theme';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';

import { Button } from '../Button';
import { Checkbox } from '../Checkbox';
import { Dropdown } from '../Dropdown';

import { ButtonGroup } from './index';

test('renders ButtonGroup', () => {
  const { container } = render(<ButtonGroup />);

  expect(container.firstChild).toMatchSnapshot();
});

test('renders ButtonGroup with Buttons', () => {
  const { container } = render(
    <ButtonGroup>
      <Button>Test</Button>
      <Button>Test</Button>
      <Button>Test</Button>
    </ButtonGroup>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('renders middle Button with no border-radius', () => {
  const testId = 'middleButton';

  const { queryByTestId } = render(
    <ButtonGroup>
      <Button>Test</Button>
      <Button data-testid={testId}>Test</Button>
      <Button>Test</Button>
    </ButtonGroup>,
  );

  expect(queryByTestId(testId)).toHaveStyle('border-radius: 0');
});

test("renders Buttons with appropriate border-radius'", () => {
  const firstTestId = 'firstButton';
  const lastTestId = 'lastButton';

  const { queryByTestId } = render(
    <ButtonGroup>
      <Button data-testid={firstTestId}>Test</Button>
      <Button>Test</Button>
      <Button data-testid={lastTestId}>Test</Button>
    </ButtonGroup>,
  );
  expect(queryByTestId(firstTestId)).toHaveStyle(`border-bottom-left-radius: ${theme.borderRadius.normal}`);
  expect(queryByTestId(firstTestId)).toHaveStyle(`border-top-left-radius: ${theme.borderRadius.normal}`);

  expect(queryByTestId(lastTestId)).toHaveStyle(`border-bottom-right-radius: ${theme.borderRadius.normal}`);
  expect(queryByTestId(lastTestId)).toHaveStyle(`border-top-right-radius: ${theme.borderRadius.normal}`);
});

test('renders with Dropdown', () => {
  const { queryByTestId } = render(
    <ButtonGroup>
      <Button>Button</Button>
      <Button>Button</Button>
      <Dropdown data-testid="dropdown" trigger={<Button data-testid="dropdown-button">Button</Button>}>
        <Dropdown.Item>Dropdown Item</Dropdown.Item>
        <Dropdown.Item>Dropdown Item</Dropdown.Item>
        <Dropdown.Item>Dropdown Item</Dropdown.Item>
      </Dropdown>
    </ButtonGroup>,
  );

  expect(queryByTestId('dropdown')).toBeInTheDocument();
  expect(queryByTestId('dropdown-button')).toBeInTheDocument();
});

test("doesn't render with invalid Components", () => {
  const checkbox = <Checkbox label="Test" />;

  const { queryByLabelText } = render(<ButtonGroup>{checkbox}</ButtonGroup>);

  expect(queryByLabelText('Test')).not.toBeInTheDocument();
});

test('does not forward styles', () => {
  const { container } = render(<ButtonGroup className="test" style={{ background: 'red' }} />);

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});
