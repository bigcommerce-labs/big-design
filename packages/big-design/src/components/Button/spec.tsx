import 'jest-styled-components';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';

import { PlusIcon } from '../Icons';

import { Button } from './index';
import { StyleableButton } from './private';

test('render default button', () => {
  const { container } = render(<Button>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render disabled button', () => {
  const { container } = render(<Button disabled>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render destructive button', () => {
  const { container } = render(<Button actionType="destructive">Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render destructive disabled button', () => {
  const { container } = render(
    <Button actionType="destructive" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary button', () => {
  const { container } = render(<Button variant="secondary">Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary disabled button', () => {
  const { container } = render(
    <Button variant="secondary" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary destructive button', () => {
  const { container } = render(
    <Button variant="secondary" actionType="destructive">
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render secondary destructive disabled button', () => {
  const { container } = render(
    <Button variant="secondary" actionType="destructive" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle button', () => {
  const { container } = render(<Button variant="subtle">Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle disabled button', () => {
  const { container } = render(
    <Button variant="subtle" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle destructive button', () => {
  const { container } = render(
    <Button variant="subtle" actionType="destructive">
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render subtle destructive disabled button', () => {
  const { container } = render(
    <Button variant="subtle" actionType="destructive" disabled>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render loading button', () => {
  const { container } = render(<Button isLoading={true}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon only button', () => {
  const { container } = render(<Button iconOnly={<PlusIcon />}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon left button', () => {
  const { container } = render(<Button iconLeft={<PlusIcon />}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon right button', () => {
  const { container } = render(<Button iconRight={<PlusIcon />}>Button</Button>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render icon left and right button', () => {
  const { container } = render(
    <Button iconLeft={<PlusIcon />} iconRight={<PlusIcon />}>
      Button
    </Button>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('forwards ref', () => {
  const ref = React.createRef<HTMLButtonElement>();

  const { container } = render(<Button ref={ref} />);
  const button = container.querySelector('button');

  expect(button).toBe(ref.current);
});

test('triggers onClick', () => {
  const onClick = jest.fn();

  const { container } = render(<Button onClick={onClick} />);
  const button = container.firstChild as HTMLElement;

  fireEvent.click(button);

  expect(onClick).toHaveBeenCalled();
});

test('does not forward styles', () => {
  const { container } = render(<Button className="test" style={{ background: 'red' }} />);

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});

test('private StyleableButton forwards styles', () => {
  const { container } = render(<StyleableButton className="test" style={{ background: 'red' }} />);

  expect(container.getElementsByClassName('test').length).toBe(1);
  expect(container.firstChild).toHaveStyle('background: red');
});

test('render only icon only with left and right icons button', () => {
  const plusIcon = <PlusIcon data-testid="icon-only" />;
  const { getAllByTestId } = render(
    <Button iconLeft={plusIcon} iconOnly={plusIcon} iconRight={plusIcon}>
      Button
    </Button>,
  );

  expect(getAllByTestId('icon-only').length).toBe(1);
});
