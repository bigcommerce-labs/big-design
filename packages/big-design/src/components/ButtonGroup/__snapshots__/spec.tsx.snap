// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ButtonGroup 1`] = `
.c2 {
  box-sizing: border-box;
}

.c1 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.bd-button + .c0 {
  margin-left: 0.5rem;
}

.c0 + .bd-button {
  margin-left: 0.5rem;
}

.c0 .bd-button {
  border-radius: 0;
  margin: 0;
  width: auto;
}

.c0 .bd-button:not(:first-of-type) {
  border-left: none;
}

.c0 .bd-button:first-of-type {
  border-bottom-left-radius: 0.25rem;
  border-top-left-radius: 0.25rem;
}

.c0 .bd-button:last-of-type {
  border-bottom-right-radius: 0.25rem;
  border-top-right-radius: 0.25rem;
}

<div
  class="c0 c1 c2"
/>
`;

exports[`renders ButtonGroup with Buttons 1`] = `
.c2 {
  box-sizing: border-box;
}

.c1 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  pointer-events: auto;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: #3C64F4;
  border-color: #3C64F4;
  font-weight: 600;
}

.c3:focus {
  outline: none;
}

.c3[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c3 + .bd-button {
  margin-top: 0.5rem;
}

.c3:active {
  background-color: #0B38D9;
}

.c3:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c3:hover:not(:active) {
  background-color: #2852EB;
}

.c3[disabled] {
  background-color: #D9DCE9;
}

.c4 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
  visibility: visible;
}

.bd-button + .c0 {
  margin-left: 0.5rem;
}

.c0 + .bd-button {
  margin-left: 0.5rem;
}

.c0 .bd-button {
  border-radius: 0;
  margin: 0;
  width: auto;
}

.c0 .bd-button:not(:first-of-type) {
  border-left: none;
}

.c0 .bd-button:first-of-type {
  border-bottom-left-radius: 0.25rem;
  border-top-left-radius: 0.25rem;
}

.c0 .bd-button:last-of-type {
  border-bottom-right-radius: 0.25rem;
  border-top-right-radius: 0.25rem;
}

@media (min-width:720px) {
  .c3 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c3 {
    width: auto;
  }
}

<div
  class="c0 c1 c2"
>
  <button
    class="bd-button c3"
    role="button"
    tabindex="0"
  >
    <span
      class="c4"
    >
      Test
    </span>
  </button>
  <button
    class="bd-button c3"
    role="button"
    tabindex="0"
  >
    <span
      class="c4"
    >
      Test
    </span>
  </button>
  <button
    class="bd-button c3"
    role="button"
    tabindex="0"
  >
    <span
      class="c4"
    >
      Test
    </span>
  </button>
</div>
`;
